src/deprecated.jl

```diff
@@ -105,3 +105,40 @@ function categorical!(df::DataFrame, cols::Union{Type, Nothing}=nothing;
     end
     return transform!(df, names(df, cols) .=> (x -> categorical(x, compress=compress)), renamecols=false)
 end
+
+@deprecate DataFrame(pairs::NTuple{N, Pair}; makeunique::Bool=false,
+          copycols::Bool=true) where {N} DataFrame(pairs..., makeunique=makeunique, copycols=copycols)
+@deprecate DataFrame(columns::NTuple{N, AbstractVector}, cnames::NTuple{N, Symbol}; makeunique::Bool=false,
+          copycols::Bool=true) where {N} DataFrame(collect(AbstractVector, columns), collect(Symbol, cnames);
+              makeunique=makeunique, copycols=copycols)
+@deprecate DataFrame(columns::NTuple{N, AbstractVector}, cnames::NTuple{N, AbstractString}; makeunique::Bool=false,
+                     copycols::Bool=true) where {N} DataFrame(collect(AbstractVector, columns), [Symbol(c) for c in cnames];
+                                                              makeunique=makeunique, copycols=copycols)
+@deprecate DataFrame(columns::NTuple{N, AbstractVector};
+                     copycols::Bool=true) where {N} DataFrame(collect(AbstractVector, columns),
+                                                              gennames(length(columns)), copycols=copycols)
+@deprecate DataFrame(columns::AbstractMatrix, cnames::AbstractVector{Symbol} = gennames(size(columns, 2));
+                     makeunique::Bool=false) DataFrame(AbstractVector[columns[:, i] for i in 1:size(columns, 2)],
+                                                       cnames; makeunique=makeunique, copycols=false)
+
+@deprecate DataFrame(columns::AbstractMatrix, cnames::AbstractVector{<:AbstractString};
+                     makeunique::Bool=false) DataFrame(AbstractVector[columns[:, i] for i in 1:size(columns, 2)],
+                                                       Symbol.(cnames); makeunique=makeunique, copycols=false)
+
+function DataFrame(column_eltypes::AbstractVector{T}, cnames::AbstractVector{Symbol},
```

Comment:
I'd be tempted to keep this one. It doesn't seem to make any harm, and the replacement is quite long if you want to allocate a nonzero number of rows. (Sorry for not commenting before.)

Reply:
I am largely indifferent. @oxinabox wanted to remove them (but maybe it was a mild preference?)

Reply:
> It doesn't seem to make any harm

Every additional way to do the same thing makes it harder to learn to do the thing.
This one is just very out of place vs the others.
A particular problem is if people find this one first and think it is the normal one.
(In-fact, i think something like that happened to me years ago when i was first learning julia.
The only way to construct a `DataFrame` i knew was the `DataFrame(cnames, ctypes)` and I kept writing awkward code to use it.
Eventually this resulted in https://github.com/JuliaData/DataFrames.jl/pull/1416)
I might not have the history quite right, but I know I was doing something awkward and was confused by constructors.


This operation seems really obscure to me.
I am not sure exactly what circumstances  you encounter this, but the one i can kind of imagine is:
where `column_eltypes` and `cnames` and `nrows` are known to the author (not computed).
and for that it is clearer to write it out.
e.g for `column_eltypes=[Int, Bool, String]` and `cnames=["A","B","C"]` and `nrows=9`
```
DataFame(
    "A"=>Vector{Int}(undef, 4),
    "B"=>Vector{Bool}(undef, 4),
    "C"=>Vector{String}(undef, 4),
)
```
I feel like if you are computing `column_eltypes` and `cnames`  you are not using this constructor in the first place, are instead doing:
```
df = DataFrame()
col_names, col_types = compute_names_and_types()
for cname, ctype in zip(col_names, col_types)
    df[!, cname] = Vector{ctype}(undef, nrows)
end
```
Both of my code examples here seem cleaner to read than the
`DataFrame(cnames, ctypes; nrows)` constructor.

Reply:
@nalimilan - I am leaning towards removing them. The reason is that usually doing:
```
DataFame(
    "A"=>Vector{Int}(undef, 4),
    "B"=>Vector{Bool}(undef, 4),
    "C"=>Vector{String}(undef, 4),
)
```
or similar is a bad idea, as most likely later the user will start doing inefficient operations.

So as @oxinabox said:
1. experienced users will not cry when they are gone
2. new users - might think it is something that is recommended, while actually it should be discouraged (one should use other patterns, like: `push!`, `append!`, transformation functions etc. to populate columns; as likely if someone writes such code the next thing will be a loop over rows with `df[i,j] = something` inside)

Reply:
OK, I'm not willing to die on that hill. Though should we really recommend using `push!` over allocating the full data frame upfront? The latter will generally overallocate.

Reply:
Yes this is true that `push!` will be slower than preallocating and iterating.

@pdeffebach, @matthieugomez, @quinnj - can you comment if you think we should keep or drop the `eltype` constructor?

I have never (except for writing tests 😄) needed to use this constructor. And in cases I need an empty data frame I always prefer something like `DataFrame(a=Int[], b=Float64[])` as it reads more easily than `DataFrame([Int, Float64], [:a, :b])` as you clearly see which column gets what type.

Reply:
@matthieugomez - is :+1: for dropping it? (just to be sure as I tried to be neutral with the question)

Reply:
FWIW: I agree with oxinabox!

Reply:
I use the `DataFrame(::AbstractVector{DataType}, ::AbstractVector{<:Union{Symbol,String}}, ::Integer)` constructor quite a bit. I am a bit conflicted... Would have to update quite a bit of codebase but if convinced it helps I am okay with it. However, maybe I can force myself to write it more descriptive. So consider myself convinced and +1 on it.